xwiki-platform-core/xwiki-platform-oldcore/src/test/java/com/xpn/xwiki/doc/XWikiDocumentMockitoTest.java

```diff
@@ -145,4 +163,169 @@ public void setTranslationAsContextDoc() throws Exception
+     */
+    private Map<String, String[]> generateFakeRequestMap()
+    {
+        Map<String, String[]> parameters = new HashMap<>();
+        // Testing update of values in existing object with existing properties
+        String[] string1 = {"string1"};
+        parameters.put("space.page_0_string", string1);
+        String[] int1 = {"7"};
+        parameters.put("space.page_1_int", int1);
+        // Testing creation and update of an object's properties when object
+        // doesn't exist
+        String[] string2 = {"string2"};
+        String[] int2 = {"13"};
+        parameters.put("space.page_2_string", string2);
+        parameters.put("space.page_2_int", int2);
+        // Testing that objects with non-following number is not created
+        parameters.put("space.page_42_string", string1);
+        parameters.put("space.page_42_int", int1);
+        // Testing that invalid parameter are ignored
+        parameters.put("invalid", new String[] {"whatever"});
+        // Testing that invalid xclass page are ignored
+        parameters.put("InvalidSpace.InvalidPage_0_string", new String[] {"whatever"});
+        // Testing that an invalid number is ignored (first should be ignored by
+        // regexp parser, second by an exception)
+        parameters.put("space.page_notANumber_string", new String[] {"whatever"});
+        parameters.put("space.page_9999999999_string", new String[] {"whatever"});
+        return parameters;
+    }
+
+    /**
+     * Generate the fake class that is used for the test of {@link #readObjectsFromForm()} and
+     * {@link #readObjectsFromFormUpdateOrCreate()}.
+     * 
+     * @return The fake BaseClass
+     */
+    private BaseClass generateFakeClass()
+    {
+        BaseClass baseClass = this.document.getXClass();
+        baseClass.addTextField("string", "String", 30);
+        baseClass.addTextAreaField("area", "Area", 10, 10);
+        baseClass.addTextAreaField("puretextarea", "Pure text area", 10, 10);
+        // set the text areas an non interpreted content
+        ((TextAreaClass) baseClass.getField("puretextarea")).setContentType("puretext");
+        baseClass.addPasswordField("passwd", "Password", 30);
+        baseClass.addBooleanField("boolean", "Boolean", "yesno");
+        baseClass.addNumberField("int", "Int", 10, "integer");
+        baseClass.addStaticListField("stringlist", "StringList", "value1, value2");
+
+        return baseClass;
+    }
+
+    /**
+     * Generate 2 clones of a fake object in the document
+     * 
+     * @return Return the reference of the first clone
+     */
+    private void generateFakeObjects()
+    {
+        BaseObject baseObject = null, baseObject2 = null;
+        try {
+            baseObject = this.document.newXObject(this.document.getDocumentReference(), this.oldcore.getXWikiContext());
+            baseObject2 =
+                this.document.newXObject(this.document.getDocumentReference(), this.oldcore.getXWikiContext());
```

Comment:
You probably need to add more objects (right now you have index 0 and 1 already existing and overridden by the request) so that you get to index 3 for example, that the request does not touch and you need to ensure that this object is still there after the request is processed (i.e. object with index 3 that was previously existing is not removed by the request processing).

...basically test the case I mentioned above when you were doing ` getXObjects().put(requestClassReference, newObjects);`